File: FourSum/simple.py
# A simple Python 3 program
# to find four elements whose
# sum is equal to zero
 
# Prints True if found False if unfound

### Problem needs to fix: it didn't pass A_50weed1.txt. Outcome should be: True but it turned out False
 
 
from __future__ import print_function
import sys
 
N = int(sys.stdin.readline()) # Provided by teacher, don't change
vals = list(map(int, sys.stdin.readlines())) # Provided by teacher, don't change
if len(vals) <= 3:
    print(False)
else:
    for i in range(0, N - 3):
        for j in range(i + 1, N - 2):
            for k in range(j + 1, N - 1):
                for l in range(k + 1, N):
                    if vals[i]+vals[j]+vals[k]+vals[l] == 0: # Provided by teacher, don't change
                        print(i, j, k, l, file=sys.stderr) # Provided by teacher, don't change
                        print(True) # Provided by teacher, don't change
                        sys.exit() # Provided by teacher, don't change
                    elif vals[i] + vals[j] + vals[k] + vals[l] < 0:
                        k = k + 1
                    else:
                        l = l - 1
                        print(False)
                        sys.exit()

File: FourSum/Simple.java
// Requirement: Find in a given array 4 numbers which sum up to 0
// Problem: Test with input Bfalse.txt, get wrong outcome. Expected outcome should be False but mine is True. 
// Is that because Bfalse.txt contains duplicates(2 times of number 4)?
// Can you implement Binary Search in Java for this assignment:?

import java.util.Scanner;
import java.util.Arrays;

public class FourSum {

    // print distinct 4-tuples (i, j, k, l) such that a[i] + a[j] + a[k] + a[l] = 0

    public static void main(String[] args) { // Provided by teacher, don't change
        Scanner S = new Scanner(System.in); // Provided by teacher, don't change
        int n = Integer.parseInt(S.nextLine()); // Provided by teacher, don't change
        long[] a = new long[n]; // Provided by teacher, don't change

        for (int i = 0; i < n; i++)
        {
            a[i] = S.nextLong();
        }

        if ( a.length < 4) {
            System.out.println("False");
        } else {
            Arrays.sort(a);
            for (int i = 0; i < n-3; i++) {
                for (int j = i + 1; j < n-2; j++) {
                    for (int k = j + 1; k < n-1; k++) {
                        for (int l = k + 1; l < n; l++) {
                            long sum = a[i] + a[j] + a[k] + a[l];
                            if (sum == 0) {
                                System.err.println(a[i] + " " + a[j] + " " + a[k] + " " + a[l]); // Provided by teacher, don't change
                                System.out.println("True"); // Provided by teacher, don't change
                                System.exit(0); // Provided by teacher, don't change
                            }
                        }
                    }
                }
            }
            System.out.println("False");
        }
    }
}
